tests/ga/extension_emulator.py

```diff
@@ -0,0 +1,315 @@
+                    raise Exception("Unexpected invocation: have ({0}, {1}, {2}), but expected ({3}, {4}, {5})".format(
+                        ext_name, ext_ver, ext_cmd, ext.name, ext.version, command_name
+                    ))
+
+            except IndexError:
+                raise Exception("No more invocations recorded. Expected ({0}, {1}, {2}).".format(ext.name, ext.version, command_name))
+        
+        if self._queue:
+            raise Exception("Invocation recorded, but not expected: ({0}, {1}, {2})".format(
+                *self._queue[0]
+            ))
+
+class ExtensionEmulator:
+    """
+    A wrapper class for the possible actions and options that an extension might support.
+    """
+
+    def __init__(self, name, version,
+        update_mode, report_heartbeat,
+        continue_on_update_failure,
+        install_action, uninstall_action,
+        enable_action, disable_action,
+        update_action):
+
+        self.name = name
+        self.version = version
+
+        self.update_mode = update_mode
+        self.report_heartbeat = report_heartbeat
+        self.continue_on_update_failure = continue_on_update_failure
+
+        self.actions = {
+            ExtensionCommandNames.INSTALL: install_action,
+            ExtensionCommandNames.UNINSTALL: uninstall_action,
+            ExtensionCommandNames.UPDATE: update_action,
+            ExtensionCommandNames.ENABLE: enable_action,
+            ExtensionCommandNames.DISABLE: disable_action
+        }
+
+        self.status_blobs = []
+        
+    
+    def matches(self, name, version):
+        return self.name == name and self.version == version
+
+def generate_patched_popen(invocation_record, *emulators):
+    """
+    Create a mock popen function able to invoke the proper action for an extension
+    emulator in emulators.
+    """
+    original_popen = subprocess.Popen
+
+    def patched_popen(cmd, *args, **kwargs):
+
+        try:
+            ext_name, ext_version, command_name = _ExtractExtensionInfo.from_command(cmd)
+            invocation_record.add(ext_name, ext_version, ExtensionCommandNames(command_name))
+        except ValueError:
+            return original_popen(cmd, *args, **kwargs)
+        
+        try:
+            return next(
+                emulator.actions[ExtensionCommandNames(command_name)]
+                for emulator in emulators
+                if emulator.matches(ext_name, ext_version)
+            )(cmd, *args, **kwargs)
+
+        except StopIteration:
+            raise Exception("Extension('{name}', '{version}') not listed as a parameter. Is it being emulated?".format(
+                name=ext_name, version=ext_version
+            ))
+
+    return patched_popen
+
+def generate_mock_load_manifest(*emulators):
+
+    original_load_manifest = ExtHandlerInstance.load_manifest
+
+    def mock_load_manifest(self):
+
+        try:
+            matching_emulator = next(
```

Comment:
I can see how this could lead to clarity issues. Your assumption is correct: `next` returns the "next" element in a list, which will always be the first matching emulator-- subsequent calls will not iterate the same list.

I personally think a better name for this function would be `first`, but I guess whomever wrote the builtins disagreed... a helper function would certainly help with this.